### PR TITLE
Build inlined functions in pprof

### PR DESCRIPTION
**Pull request by piotmag769**
Related #71 

## Introduced changes

- Take inlined functions into account when building pprof profile

### CODE REVIEW

File: crates/cairo-profiler/src/profile_builder.rs [hunk 1]
```diff
@@ -66,32 +66,74 @@ impl ProfilerContext {
         }
     }
 
-    fn location_id(&mut self, location: &Function) -> LocationId {
-        let location = match location {
-            Function::Entrypoint(function_name) | Function::NonInlined(function_name) => {
-                function_name
+    fn locations_ids(&mut self, call_stack: &[Function]) -> Vec<LocationId> {
+        let mut locations_ids = vec![];
+        // This vector represent stacks of functions corresponding to single locations.
```
Comment (piotmag769) [inline crates/cairo-profiler/src/profile_builder.rs:71]: This is bcs in pprof inlining is signalised by locations having multiple lines. So e.g. stack like that
`[A, B_inlined, C, D, E_inlined, F_inlined]` will be converted to 3 pprof locations corresponding to the following stacks:
- `[A, B_inlined]`
- `[C]`
- `[D, E_inlined, F_inlined]`

File: crates/cairo-profiler/src/profile_builder.rs [hunk 2]
```diff
@@ -66,32 +66,74 @@ impl ProfilerContext {
...
+                        };
+                         pprof::Location {
+                            id: (self.locations.len() + 1) as u64,
+                            mapping_id: 0,
+                            address: 0,
+                            line: vec![line],
+                            is_folded: true,
+                        }
+                    }
+                    Function::Inlined(_) => unreachable!("First function in a function stack corresponding to a single location cannot be inlined")
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-profiler/src/profile_builder.rs:110]:
```suggestion
                    Function::Inlined(_) => unreachable!("First function in a call stack corresponding to a single location cannot be inlined")
```

File: crates/cairo-profiler/src/profile_builder.rs [hunk 3]
```diff
@@ -66,32 +66,74 @@ impl ProfilerContext {
         }
     }
 
-    fn location_id(&mut self, location: &Function) -> LocationId {
-        let location = match location {
-            Function::Entrypoint(function_name) | Function::NonInlined(function_name) => {
-                function_name
+    fn locations_ids(&mut self, call_stack: &[Function]) -> Vec<LocationId> {
+        let mut locations_ids = vec![];
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-profiler/src/profile_builder.rs:70]: No logic should be performed in `profile_builder` you should do this in previous layer and mark elements in a sample location as inlined.

File: crates/cairo-profiler/src/trace_reader/sample.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
+use crate::trace_reader::function_name::FunctionName;
```
Comment (piotmag769) [inline crates/cairo-profiler/src/trace_reader/sample.rs:1]: Just moved to another file, the previous one was getting big

File: crates/cairo-profiler/src/profile_builder.rs (hunk 3 above)
Comment (piotmag769) [reply crates/cairo-profiler/src/profile_builder.rs:72]: Done

File: crates/cairo-profiler/src/trace_reader.rs [hunk 5]
```diff
@@ -207,3 +122,45 @@ fn collect_samples<'a>(
...
+        current_function_stack_start_index,
+        sample.call_stack.len() - 1,
+    ));
+
+    let mut aggregated_call_stack = vec![];
+    let call_stack_iter = sample.call_stack.into_iter();
+    for (start_index, end_index) in function_stacks_indexes {
+        aggregated_call_stack.push(
+            call_stack_iter
+                .clone()
```
Comment (piotmag769) [inline crates/cairo-profiler/src/trace_reader.rs:156]: Cloning an iterator is rly cheat as it doesn't clone new elements

File: crates/cairo-profiler/src/trace_reader/sample.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+use std::collections::HashMap;
+use trace_data::{ExecutionResources, L1Resources};
+
+pub struct AggregatedSample {
+    pub aggregated_call_stack: Vec<Vec<Function>>,
+    pub measurements: HashMap<MeasurementUnit, MeasurementValue>,
+}
```
Comment (piotmag769) [inline crates/cairo-profiler/src/trace_reader/sample.rs:8]: Only this struct is new here

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Comment by MaksymilianDemitraszek** [comment]
Please rol back the commits before merging

**Pull request merged**